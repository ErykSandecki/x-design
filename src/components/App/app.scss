.App {
  @include theme('neutral-4', background-color);
  @include theme('neutral-1', color);

  min-height: 100%;
  overflow: auto;
}
