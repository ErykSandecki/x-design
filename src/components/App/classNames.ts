// utils
import { composeClassNames } from 'utils';

export const className = 'App';

export const classNames = composeClassNames(className);
