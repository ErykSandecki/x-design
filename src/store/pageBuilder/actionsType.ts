export const REDUCER_KEY = 'pageBuilder';

export const ADD_ELEMENT = `${REDUCER_KEY}/ADD_ELEMENT`;
export const APPLY_ELEMENTS_SIZE_TYPE = `${REDUCER_KEY}/APPLY_ELEMENTS_SIZE_TYPE`;
export const CHANGE_ALIGNMENT = `${REDUCER_KEY}/CHANGE_ALIGNMENT`;
export const CHANGE_BACKGROUND = `${REDUCER_KEY}/CHANGE_BACKGROUND`;
export const CHANGE_LAYOUT = `${REDUCER_KEY}/CHANGE_LAYOUT`;
export const CHANGE_PARENT = `${REDUCER_KEY}/CHANGE_PARENT`;
export const CHANGE_POSITION = `${REDUCER_KEY}/CHANGE_POSITION`;
export const CLEAR_PREV_STATE = `${REDUCER_KEY}/CLEAR_PREV_STATE`;
export const FIT_LAYOUT = `${REDUCER_KEY}/FIT_LAYOUT`;
export const FLIP_ELEMENTS = `${REDUCER_KEY}/FLIP_ELEMENTS`;
export const REDUCER_HISTORY_REDO = `${REDUCER_KEY}/REDUCER_HISTORY_REDO`;
export const REDUCER_HISTORY_SAVE = `${REDUCER_KEY}/REDUCER_HISTORY_SAVE`;
export const REDUCER_HISTORY_UNDO = `${REDUCER_KEY}/REDUCER_HISTORY_UNDO`;
export const RESIZE_ELEMENT = `${REDUCER_KEY}/RESIZE_ELEMENT`;
export const ROTATE_ELEMENTS = `${REDUCER_KEY}/ROTATE_ELEMENTS`;
export const SELECT_ELEMENT = `${REDUCER_KEY}/SELECT_ELEMENT`;
export const SELECT_ELEMENTS = `${REDUCER_KEY}/SELECT_ELEMENTS`;
export const SET_AREA_COORDINATES = `${REDUCER_KEY}/SET_AREA_COORDINATES`;
export const SET_ELEMENTS_COORDINATES = `${REDUCER_KEY}/SET_ELEMENTS_COORDINATES`;
export const SET_ELEMENTS_SCORE_TO_CURRENT_SIZE = `${REDUCER_KEY}/SET_ELEMENTS_SCORE_TO_CURRENT_SIZE`;
export const SET_ELEMENTS_SIZES = `${REDUCER_KEY}/SET_ELEMENTS_SIZES`;
export const SET_ELEMENTS_SIZES_MIN_MAX = `${REDUCER_KEY}/SET_ELEMENTS_SIZES_MIN_MAX`;
export const TOGGLE_ASPECT_RATIO = `${REDUCER_KEY}/TOGGLE_ASPECT_RATIO`;
export const UPDATE_EVENTS_STATUS = `${REDUCER_KEY}/UPDATE_EVENTS_STATUS`;
export const UPDATE_PREV_STATE = `${REDUCER_KEY}/UPDATE_PREV_STATE`;
export const UNSELECT_ELEMENT = `${REDUCER_KEY}/UNSELECT_ELEMENT`;
