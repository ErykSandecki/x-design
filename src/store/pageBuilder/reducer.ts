// others
import {
  ADD_ELEMENT,
  APPLY_ELEMENTS_INSET_TYPE,
  APPLY_ELEMENTS_SIZE_MIN_MAX_TYPE,
  CHANGE_ALIGNMENT,
  CHANGE_BACKGROUND,
  CHANGE_LAYOUT,
  CHANGE_PARENT,
  CHANGE_POSITION,
  CLEAR_PREV_STATE,
  CHANGE_LAYOUT_BOX_SIZING,
  CHANGE_LAYOUT_GRID,
  CHANGE_INSETS,
  FIT_LAYOUT,
  FLIP_ELEMENTS,
  REDUCER_HISTORY_REDO,
  REDUCER_HISTORY_SAVE,
  REDUCER_HISTORY_UNDO,
  ROTATE_ELEMENTS,
  SELECT_ELEMENT,
  SELECT_ELEMENTS,
  SET_AREA_COORDINATES,
  RESIZE_ELEMENT,
  SET_ELEMENTS_COORDINATES,
  UNSELECT_ELEMENT,
  UPDATE_EVENTS_STATUS,
  UPDATE_PREV_STATE,
  SET_ELEMENTS_SIZES,
  APPLY_ELEMENTS_SIZE_TYPE,
  SET_ELEMENTS_SIZES_MIN_MAX,
  SET_ELEMENTS_SCORE_TO_CURRENT_SIZE,
  TOGGLE_ASPECT_RATIO,
  CHANGE_LAYOUT_ALIGNMENT,
  SET_ELEMENTS_GAP,
  APPLY_ELEMENTS_GAP_TYPE,
  CHANGE_CLIP_CONTENT,
} from './actionsType';
import { BASE_PAGE } from './constants';

// types
import { AnchorResize, AnchorRotate } from './enums';
import { KeyboardKeys, TAction } from 'types';
import {
  TAddELementAction,
  TSelectElementAction,
  TPageBuilderState,
  TUnselectElementAction,
  TSelectElementsAction,
  TSetAreaCoordinatesAction,
  TUpdateEventsStatusAction,
  TResizeElementAction,
  TRotateElementsAction,
  TChangeBackgroundAction,
  TReducerHistorySaveAction,
  TChangeAlignmentAction,
  TFlipElementsAction,
  TChangeLayoutAction,
  TSetElementsCoordinatesAction,
  TSetElementsSizesAction,
  TApplyElementsSizeTypeAction,
  TSetElementsSizesMinMaxAction,
  TSetElementsScoreToCurrentSizeAction,
  TChangeLayoutAlignmentAction,
  TSetElementsGapAction,
  TApplyElementsGapTypeAction,
  TChangeLayoutBoxSizingAction,
  TChangeLayoutGridAction,
  TChangeInsetsAction,
  TApplyElementsInsetTypeAction,
  TApplyElementsSizeMinMaxTypeAction,
  TChangeClipContentAction,
} from './types';

// utils
import { filterSelectedElements } from './utils/filterSelectedElements';
import { handleAddElement } from './utils/handleAddElement';
import { handleApplyElementsGapType } from './utils/applyElementsGapType/handleApplyElementsGapType';
import { handleApplyElementsInsetType } from './utils/applyElementsInsetType/handleApplyElementsInsetType';
import { handleApplyElementsSizeMinMaxType } from './utils/applyElementsSizeMinMaxType/handleApplyElementsSizeMinMaxType';
import { handleApplyElementsSizeType } from './utils/applyElementsSizeType/handleApplyElementsSizeType';
import { handleChangeAlignment } from './utils/changeAligment/handleChangeAlignment';
import { handleChangeBackground } from './utils/handleChangeBackground';
import { handleChangeClipContent } from './utils/handleChangeClipContent';
import { handleChangeInsets } from './utils/handleChangeInsets';
import { handleChangeLayout } from './utils/changeLayout/handleChangeLayout';
import { handleChangeLayoutAlignment } from './utils/handleChangeLayoutAlignment';
import { handleChangeLayoutBoxSizing } from './utils/handleChangeLayoutBoxSizing';
import { handleChangeLayoutGrid } from './utils/changeLayoutGrid/handleChangeLayoutGrid';
import { handleChangeParent } from './utils/changeParent/handleChangeParent';
import { handleChangePosition } from './utils/handleChangePosition';
import { handleFitLayout } from './utils/handleFitLayout';
import { handleFlipElements } from './utils/handleFlipElements';
import { handleReducerHistoryRedo } from './utils/reducerHistory/handleReducerHistoryRedo';
import { handleReducerHistorySave } from './utils/reducerHistory/handleReducerHistorySave';
import { handleReducerHistoryUndo } from './utils/reducerHistory/handleReducerHistoryUndo';
import { handleResizeElement } from './utils/resizeElement/handleResizeElement';
import { handleRotateElements } from './utils/handleRotateElements';
import { handleSetElementsCoordinates } from './utils/handleSetElementsCoordinates';
import { handleSetElementsGap } from './utils/handleSetElementsGap';
import { handleSetElementsScoreToCurrentSize } from './utils/handleSetElementsScoreToCurrentSize';
import { handleSetElementsSizes } from './utils/setElementSizes/handleSetElementsSizes';
import { handleSetElementsSizesMinMax } from './utils/handleSetElementsSizesMinMax';
import { handleToggleAspectRatio } from './utils/handleToggleAspectRatio';

const initialState: TPageBuilderState = {
  currentPage: '0',
  events: {
    canMoveElements: true,
    colorSampler: false,
    draggableElements: [],
    hoverOnElement: '-1',
    isGridDropArea: false,
    isMultipleMoving: false,
    isResizing: false,
    isRotating: false,
    possibleAnchorElementId: '-1',
    possibleAnchorPosition: null,
    possibleElement: undefined,
    possibleIndexPosition: null,
    possibleParent: null,
    pressedKey: KeyboardKeys.none,
    selectedAnchorResize: AnchorResize.none,
    selectedAnchorRotate: AnchorRotate.none,
  },
  isLoading: true,
  isPending: false,
  // @ts-ignore
  pages: {
    [BASE_PAGE.id]: {
      ...BASE_PAGE,
    },
  },
};

const addElement = (
  state: TPageBuilderState,
  { payload: element }: TAction<TAddELementAction['payload']>,
): TPageBuilderState => handleAddElement(element, state);

const applyElementsGapType = (
  state: TPageBuilderState,
  { payload }: TAction<TApplyElementsGapTypeAction['payload']>,
): TPageBuilderState => handleApplyElementsGapType(payload, state);

const applyElementsInsetType = (
  state: TPageBuilderState,
  { payload }: TAction<TApplyElementsInsetTypeAction['payload']>,
): TPageBuilderState => handleApplyElementsInsetType(payload, state);

const applyElementsSizeMinMaxType = (
  state: TPageBuilderState,
  { payload }: TAction<TApplyElementsSizeMinMaxTypeAction['payload']>,
): TPageBuilderState => handleApplyElementsSizeMinMaxType(payload, state);

const applyElementsSizeType = (
  state: TPageBuilderState,
  { payload }: TAction<TApplyElementsSizeTypeAction['payload']>,
): TPageBuilderState => handleApplyElementsSizeType(payload, state);

const changeAlignment = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeAlignmentAction['payload']>,
): TPageBuilderState => handleChangeAlignment(payload, state);

const changeBackground = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeBackgroundAction['payload']>,
): TPageBuilderState => handleChangeBackground(payload, state);

const changeClipContent = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeClipContentAction['payload']>,
): TPageBuilderState => handleChangeClipContent(payload, state);

const changeInsets = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeInsetsAction['payload']>,
): TPageBuilderState => handleChangeInsets(payload, state);

const changeLayout = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeLayoutAction['payload']>,
): TPageBuilderState => handleChangeLayout(payload, state);

const changeLayoutAlignment = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeLayoutAlignmentAction['payload']>,
): TPageBuilderState => handleChangeLayoutAlignment(payload, state);

const changeLayoutBoxSizing = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeLayoutBoxSizingAction['payload']>,
): TPageBuilderState => handleChangeLayoutBoxSizing(payload, state);

const changeLayoutGrid = (
  state: TPageBuilderState,
  { payload }: TAction<TChangeLayoutGridAction['payload']>,
): TPageBuilderState => handleChangeLayoutGrid(payload, state);

const changeParent = (state: TPageBuilderState): TPageBuilderState => handleChangeParent(state);

const changePosition = (state: TPageBuilderState): TPageBuilderState => handleChangePosition(state);

const clearPrevState = (state: TPageBuilderState): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      prevState: undefined,
    },
  },
});

const fitLayout = (state: TPageBuilderState): TPageBuilderState => handleFitLayout(state);

const flipElements = (
  state: TPageBuilderState,
  { payload }: TAction<TFlipElementsAction['payload']>,
): TPageBuilderState => handleFlipElements(payload, state);

const reducerHistoryRedo = (state: TPageBuilderState): TPageBuilderState => handleReducerHistoryRedo(state);

const reducerHistorySave = (
  state: TPageBuilderState,
  { payload }: TAction<TReducerHistorySaveAction['payload']>,
): TPageBuilderState => handleReducerHistorySave(state, payload);

const reducerHistoryUndo = (state: TPageBuilderState): TPageBuilderState => handleReducerHistoryUndo(state);

const resizeElement = (
  state: TPageBuilderState,
  { payload: { baseCoordinates, flip, height, id, mouseCoordinates, width } }: TAction<TResizeElementAction['payload']>,
): TPageBuilderState => handleResizeElement(baseCoordinates, flip, height, width, id, mouseCoordinates, state);

const rotateElements = (
  state: TPageBuilderState,
  { payload }: TAction<TRotateElementsAction['payload']>,
): TPageBuilderState => handleRotateElements(payload, state);

const selectElement = (
  state: TPageBuilderState,
  { payload: selectedElement }: TAction<TSelectElementAction['payload']>,
): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      selectedElements: filterSelectedElements(
        [...state.pages[state.currentPage].selectedElements, selectedElement],
        state,
      ),
    },
  },
});

const selectElements = (
  state: TPageBuilderState,
  { payload: selectedElements }: TAction<TSelectElementsAction['payload']>,
): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      selectedElements: filterSelectedElements(selectedElements, state),
    },
  },
});

const setAreaCoordinates = (
  state: TPageBuilderState,
  { payload: areaCoordinates }: TAction<TSetAreaCoordinatesAction['payload']>,
): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      areaCoordinates: {
        ...state.pages[state.currentPage].areaCoordinates,
        ...areaCoordinates,
      },
    },
  },
});

const setElementsCoordinates = (
  state: TPageBuilderState,
  { payload: coordinates }: TAction<TSetElementsCoordinatesAction['payload']>,
): TPageBuilderState => handleSetElementsCoordinates(coordinates, state);

const setElementsGap = (
  state: TPageBuilderState,
  { payload }: TAction<TSetElementsGapAction['payload']>,
): TPageBuilderState => handleSetElementsGap(payload, state);

const setElementsScoreToCurrentSize = (
  state: TPageBuilderState,
  { payload: { scoreType, sizeType } }: TAction<TSetElementsScoreToCurrentSizeAction['payload']>,
): TPageBuilderState => handleSetElementsScoreToCurrentSize(scoreType, sizeType, state);

const setElementsSizes = (
  state: TPageBuilderState,
  { payload: { sizeType, value } }: TAction<TSetElementsSizesAction['payload']>,
): TPageBuilderState => handleSetElementsSizes(sizeType, state, value);

const setElementsSizesMinMax = (
  state: TPageBuilderState,
  { payload: { scoreType, sizeType, value } }: TAction<TSetElementsSizesMinMaxAction['payload']>,
): TPageBuilderState => handleSetElementsSizesMinMax(scoreType, sizeType, state, value);

const toggleAspectRatio = (state: TPageBuilderState): TPageBuilderState => handleToggleAspectRatio(state);

const updateEventsStatus = (
  state: TPageBuilderState,
  { payload: events }: TAction<TUpdateEventsStatusAction['payload']>,
): TPageBuilderState => ({
  ...state,
  events: {
    ...state.events,
    ...events,
  },
});

const updatePrevState = (state: TPageBuilderState): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      prevState: state.pages[state.currentPage],
    },
  },
});

const unselectElement = (
  state: TPageBuilderState,
  { payload: id }: TAction<TUnselectElementAction['payload']>,
): TPageBuilderState => ({
  ...state,
  pages: {
    ...state.pages,
    [state.currentPage]: {
      ...state.pages[state.currentPage],
      selectedElements: state.pages[state.currentPage].selectedElements.filter((element) => element.id !== id),
    },
  },
});

const pageBuilder = (state: TPageBuilderState = initialState, action: TAction): TPageBuilderState => {
  switch (action.type) {
    case ADD_ELEMENT:
      return addElement(state, action);
    case APPLY_ELEMENTS_GAP_TYPE:
      return applyElementsGapType(state, action);
    case APPLY_ELEMENTS_INSET_TYPE:
      return applyElementsInsetType(state, action);
    case APPLY_ELEMENTS_SIZE_MIN_MAX_TYPE:
      return applyElementsSizeMinMaxType(state, action);
    case APPLY_ELEMENTS_SIZE_TYPE:
      return applyElementsSizeType(state, action);
    case CHANGE_ALIGNMENT:
      return changeAlignment(state, action);
    case CHANGE_BACKGROUND:
      return changeBackground(state, action);
    case CHANGE_CLIP_CONTENT:
      return changeClipContent(state, action);
    case CHANGE_INSETS:
      return changeInsets(state, action);
    case CHANGE_LAYOUT:
      return changeLayout(state, action);
    case CHANGE_LAYOUT_ALIGNMENT:
      return changeLayoutAlignment(state, action);
    case CHANGE_LAYOUT_BOX_SIZING:
      return changeLayoutBoxSizing(state, action);
    case CHANGE_LAYOUT_GRID:
      return changeLayoutGrid(state, action);
    case CHANGE_PARENT:
      return changeParent(state);
    case CHANGE_POSITION:
      return changePosition(state);
    case CLEAR_PREV_STATE:
      return clearPrevState(state);
    case FIT_LAYOUT:
      return fitLayout(state);
    case FLIP_ELEMENTS:
      return flipElements(state, action);
    case REDUCER_HISTORY_REDO:
      return reducerHistoryRedo(state);
    case REDUCER_HISTORY_SAVE:
      return reducerHistorySave(state, action);
    case REDUCER_HISTORY_UNDO:
      return reducerHistoryUndo(state);
    case RESIZE_ELEMENT:
      return resizeElement(state, action);
    case ROTATE_ELEMENTS:
      return rotateElements(state, action);
    case SELECT_ELEMENT:
      return selectElement(state, action);
    case SELECT_ELEMENTS:
      return selectElements(state, action);
    case SET_AREA_COORDINATES:
      return setAreaCoordinates(state, action);
    case SET_ELEMENTS_COORDINATES:
      return setElementsCoordinates(state, action);
    case SET_ELEMENTS_GAP:
      return setElementsGap(state, action);
    case SET_ELEMENTS_SCORE_TO_CURRENT_SIZE:
      return setElementsScoreToCurrentSize(state, action);
    case SET_ELEMENTS_SIZES:
      return setElementsSizes(state, action);
    case SET_ELEMENTS_SIZES_MIN_MAX:
      return setElementsSizesMinMax(state, action);
    case TOGGLE_ASPECT_RATIO:
      return toggleAspectRatio(state);
    case UPDATE_EVENTS_STATUS:
      return updateEventsStatus(state, action);
    case UPDATE_PREV_STATE:
      return updatePrevState(state);
    case UNSELECT_ELEMENT:
      return unselectElement(state, action);
    default:
      return state;
  }
};

export default pageBuilder;
