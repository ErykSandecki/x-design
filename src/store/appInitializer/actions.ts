// others
import { APP_INIT, INIT_LANGUAGE, INIT_LANGUAGE_SUCCESS, SET_IS_APP_LOADED, SET_THEME } from './actionsType';

// types
import {
  TAppInitAction,
  TInitLanguageAction,
  TInitLanguageSuccessAction,
  TSetIsAppLoadedAction,
  TSetThemeAction,
} from './types';

export const appInit = (): TAppInitAction => ({
  type: APP_INIT,
});

export const initLanguage = (): TInitLanguageAction => ({
  type: INIT_LANGUAGE,
});

export const initLanguageSuccess = (payload: TInitLanguageSuccessAction['payload']): TInitLanguageSuccessAction => ({
  payload,
  type: INIT_LANGUAGE_SUCCESS,
});

export const setIsAppLoaded = (payload: TSetIsAppLoadedAction['payload']): TSetIsAppLoadedAction => ({
  payload,
  type: SET_IS_APP_LOADED,
});

export const setTheme = (payload: TSetThemeAction['payload']): TSetThemeAction => ({
  payload,
  type: SET_THEME,
});
