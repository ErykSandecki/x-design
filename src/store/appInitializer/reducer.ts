// others
import { DEFAULT_LANGUAGE } from 'translations';
import { INIT_LANGUAGE_SUCCESS, SET_IS_APP_LOADED, SET_THEME } from './actionsType';

// types
import { TAction } from 'types';
import { TAppInitializerState, TInitLanguageSuccessAction, TSetIsAppLoadedAction, TSetThemeAction } from './types';

// utils
import { getThemePreferences } from 'utils';

const initialState: TAppInitializerState = {
  isAppLoaded: false,
  isPending: true,
  language: DEFAULT_LANGUAGE,
  theme: getThemePreferences(),
};

const initLanguageSuccess = (
  state: TAppInitializerState,
  { payload: language }: TAction<TInitLanguageSuccessAction['payload']>,
): TAppInitializerState => ({
  ...state,
  language,
});

const setIsAppLoaded = (
  state: TAppInitializerState,
  { payload: isAppLoaded }: TAction<TSetIsAppLoadedAction['payload']>,
): TAppInitializerState => ({
  ...state,
  isAppLoaded,
  isPending: false,
});

const setTheme = (
  state: TAppInitializerState,
  { payload: theme }: TAction<TSetThemeAction['payload']>,
): TAppInitializerState => ({
  ...state,
  theme,
});

const appInitializer = (state: TAppInitializerState = initialState, action: TAction): TAppInitializerState => {
  switch (action.type) {
    case INIT_LANGUAGE_SUCCESS:
      return initLanguageSuccess(state, action);
    case SET_IS_APP_LOADED:
      return setIsAppLoaded(state, action);
    case SET_THEME:
      return setTheme(state, action);
    default:
      return state;
  }
};

export default appInitializer;
