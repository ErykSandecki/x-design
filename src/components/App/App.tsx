import { FC, useEffect } from 'react';
import { useSelector } from 'react-redux';

// components
import PageBuilderPage from 'pages/PageBuilderPage/PageBuilderPage';

// hooks
import { useDispatchMany } from 'hooks';

// store
import { appInit, initLanguage } from 'store/appInitializer/actions';
import { isAppLoadedSelector, isPendingSelector } from 'store/appInitializer/selectors';

const App: FC = () => {
  const dispatchMany = useDispatchMany();
  const isAppLoaded = useSelector(isAppLoadedSelector);
  const isPending = useSelector(isPendingSelector);

  useEffect(() => {
    dispatchMany(appInit(), initLanguage());
  }, []);

  if (!isAppLoaded && isPending) {
    return <main>Init</main>;
  }

  if (!isAppLoaded && !isPending) {
    return <main>Error</main>;
  }

  return <PageBuilderPage />;
};

export default App;
