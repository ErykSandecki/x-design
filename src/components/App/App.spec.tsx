import { Provider } from 'react-redux';

// components
import App from './App';

// mocks
import { wholeStateMock } from 'test/mocks/reducer/wholeStateMock';
import 'test/mocks/sagas/allSagas';

// others
import { REDUCER_KEY as APP_INITIALIZER } from 'store/appInitializer/actionsType';

// store
import { configureStore } from 'store/store';

// utils
import { customRender } from 'test';

describe('App snapshots', () => {
  it('should render initializer', () => {
    // mock
    const store = configureStore();

    // before
    const { asFragment } = customRender(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render error page', () => {
    // mock
    const store = configureStore({
      ...wholeStateMock,
      [APP_INITIALIZER]: {
        isAppLoaded: false,
        isPending: false,
      },
    });

    // before
    const { asFragment } = customRender(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render router', () => {
    // mock
    const store = configureStore({
      ...wholeStateMock,
      [APP_INITIALIZER]: {
        isAppLoaded: true,
        isPending: false,
      },
    });

    // before
    const { asFragment } = customRender(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
