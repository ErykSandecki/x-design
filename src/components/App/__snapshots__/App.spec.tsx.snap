// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshots should render error page 1`] = `
<DocumentFragment>
  <main>
    Error
  </main>
</DocumentFragment>
`;

exports[`App snapshots should render initializer 1`] = `
<DocumentFragment>
  <main>
    Init
  </main>
</DocumentFragment>
`;

exports[`App snapshots should render router 1`] = `
<DocumentFragment>
  <div
    class="css-1jnxvbp"
    data-test-box=""
  >
    <div
      class="css-1rzc7eb"
      data-test-box=""
    >
      <div
        class="css-1jnxvbp MouseModes MouseModes--light"
        data-test-box=""
      >
        <div
          class="css-1jnxvbp MouseModes__button MouseModes__button--light"
          data-test-box=""
        >
          <div
            class="css-1jnxvbp Icon Icon--light Icon--neutral1 Icon--neutral1--light"
            data-test-icon="menu"
            height="16"
            name="Logo"
            width="16"
          />
        </div>
        <div
          class="css-1jnxvbp MouseModes__button MouseModes__button--light MouseModes__button--active MouseModes__button--active--light"
          data-test-box=""
        >
          <div
            class="css-1jnxvbp Icon Icon--light Icon--neutral1 Icon--neutral1--light"
            data-test-icon="default"
            height="16"
            name="MoveTool"
            width="16"
          />
        </div>
        <div
          class="css-1jnxvbp MouseModes__button MouseModes__button--light"
          data-test-box=""
        >
          <div
            class="css-1jnxvbp Icon Icon--light Icon--neutral1 Icon--neutral1--light"
            data-test-icon="move"
            height="16"
            name="MoveHand"
            width="16"
          />
        </div>
        <div
          class="css-1jnxvbp MouseModes__button MouseModes__button--light"
          data-test-box=""
        >
          <div
            class="css-1jnxvbp Icon Icon--light Icon--neutral1 Icon--neutral1--light"
            data-test-icon="comment"
            height="16"
            name="Comment"
            width="16"
          />
        </div>
        <div
          class="css-1jnxvbp MouseModes__button MouseModes__button--light"
          data-test-box=""
        >
          <div
            class="css-1jnxvbp Icon Icon--light Icon--neutral1 Icon--neutral1--light"
            data-test-icon="toolBeltA"
            height="16"
            name="FrameTool"
            width="16"
          />
        </div>
      </div>
    </div>
    <div
      class="css-w7unf8 PanelProperties PanelProperties--light"
      data-test-box="panel-properties"
      id="panelProperties"
      style="width: 240px;"
    >
      <div
        class="PanelProperties__area-handle-resize PanelProperties__area-handle-resize--light"
        data-test-resize=""
      />
      <div
        class="css-1jnxvbp Section Section--light"
        data-test-section=""
      >
        <div
          class="Section__header Section__header--light"
        >
          <small
            class="css-1jnxvbp Section__label Section__label--light Typography Typography--light Typography--neutral1 Typography--neutral1--light Typography--regular Typography--regular--light Typography--normal Typography--normal--light Typography--small Typography--small--light"
            data-test-text=""
            style="text-align: inherit;"
          >
            Mode
          </small>
          <div
            class="Section__buttons Section__buttons--light"
          />
        </div>
        <div
          class="css-1jnxvbp Tabs Tabs--light"
          data-test-box=""
        >
          <div
            class="Tabs__tab Tabs__tab--light Tabs__tab--active Tabs__tab--active--light"
            data-test-tab="design"
          >
            Design
          </div>
          <div
            class="Tabs__tab Tabs__tab--light"
            data-test-tab="prototype\`,"
          >
            Prototype
          </div>
        </div>
      </div>
      <div
        class="PanelProperties__sections PanelProperties__sections--light"
      >
        <div
          class="css-1jnxvbp Section Section--light"
          data-test-section=""
        >
          <div
            class="Section__header Section__header--light"
          >
            <small
              class="css-1jnxvbp Section__label Section__label--light Typography Typography--light Typography--neutral1 Typography--neutral1--light Typography--regular Typography--regular--light Typography--normal Typography--normal--light Typography--small Typography--small--light"
              data-test-text=""
              style="text-align: inherit;"
            >
              Page
            </small>
            <div
              class="Section__buttons Section__buttons--light"
            />
          </div>
          <div
            class="css-1jnxvbp SectionColumn SectionColumn--light"
            data-test-section=""
          >
            <div
              class="SectionColumn__labels SectionColumn__labels--light"
              style="width: calc(100% - 24px - 0px - 8px);"
            />
            <div
              class="SectionColumn__wrapper SectionColumn__wrapper--light"
            >
              <div
                class="SectionColumn__content SectionColumn__content--light SectionColumn__content--single SectionColumn__content--single--light"
                style="width: calc(100% - 24px - 0px - 8px);"
              >
                <div
                  class="css-1jnxvbp FieldGroup FieldGroup--light"
                  data-test-field-group=""
                >
                  <div
                    class="css-1jnxvbp ColorPicker__hexInput ColorPicker__hexInput--light TextField TextField--light"
                    data-test-text-field="color"
                  >
                    <div
                      aria-describedby="test-id"
                    >
                      <div
                        class="css-1jnxvbp Color Color--light"
                        data-test-color=""
                      >
                        <div
                          class="Color__picker Color__picker--light"
                          style="background-color: rgb(241, 242, 246);"
                        />
                        <div
                          class="Color__picker-alpha Color__picker-alpha--light"
                          style="background-color: rgb(241, 242, 246);"
                        />
                        <div
                          class="Color__picker-texture Color__picker-texture--light"
                        />
                      </div>
                    </div>
                    <input
                      class="TextField__input TextField__input--light"
                      data-test-text-field-input="color"
                      maxlength="6"
                      value="f1f2f6"
                    />
                  </div>
                  <div
                    class="css-1jnxvbp ColorPicker__alphaInput ColorPicker__alphaInput--light TextField TextField--light"
                    data-test-text-field="alpha"
                  >
                    <input
                      class="TextField__input TextField__input--light"
                      data-test-text-field-input="alpha"
                      max="100"
                      maxlength="6"
                      min="0"
                      type="number"
                      value="100"
                    />
                    <div
                      class="ScrubbableInput ScrubbableInput--light"
                      data-test-scrubbable-input=""
                    >
                      <div
                        class="ColorPicker__alphaInputUnit ColorPicker__alphaInputUnit--light"
                      >
                        %
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="SectionColumn__buttons SectionColumn__buttons--light"
              >
                <div
                  class="css-1jnxvbp ButtonIcon ButtonIcon--light"
                  data-test-button-icon=""
                >
                  <div
                    class="css-1jnxvbp Icon Icon--light Icon--neutral1 Icon--neutral1--light"
                    data-test-icon="eyes-opened"
                    height="14"
                    name="EyesOpened"
                    width="14"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-117yijb ViewBox ViewBox--light ZoomBox ZoomBox--light ZoomBox--idle ZoomBox--idle--light ZoomBox--default ZoomBox--default--light"
      data-test-box="zoom-box"
      tabindex="0"
    >
      <div
        class="css-1jnxvbp ZoomBox__backgroundMask ZoomBox__backgroundMask--light"
        data-test-box=""
        style="background-color: rgb(241, 242, 246); display: initial;"
      />
      <div
        class="css-1jnxvbp ZoomBox__textureBlank ZoomBox__textureBlank--light"
        data-test-box=""
        style="transform: translate(0px, 0px) scale(1);"
      />
      <div
        class="css-1j4c14e ZoomBox__zoomContent ZoomBox__zoomContent--light"
        data-test-box=""
        id="zoomContentId"
        style="transform: translate(0px, 0px) scale(1);"
      >
        <div
          class="css-u5h2b"
          data-test-box=""
          id="overlayContainer"
          style="z-index: var(--xd-page-builder-zindex-overlay-container);"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
